File: Traditional-Important-Algorithms/exercises_py/01_gc_and_memory/gc08_write_barrier.py
# I AM NOT DONE

"""
Exercise: Write Barriers

Write barriers intercept pointer writes to maintain GC invariants.
They're essential for incremental and concurrent GC.

Your task: Implement different write barrier strategies.
"""

from enum import Enum


class Color(Enum):
    """Object color in tricolor marking"""
    WHITE = "white"
    GRAY = "gray"
    BLACK = "black"


class WriteBarrierGC:
    """GC with different write barrier implementations"""

    def __init__(self):
        self.objects = {}  # obj_id -> (color, references)
        self.next_id = 0
        self.roots = set()
        self.remembered_set = set()  # Set of (from_id, to_id) tuples

    def allocate(self, references=None):
        """Allocate a new object"""
        obj_id = self.next_id
        self.next_id += 1
        self.objects[obj_id] = (Color.WHITE, references if references else [])
        return obj_id

    def add_root(self, obj_id):
        """Add an object as a root"""
        self.roots.add(obj_id)

    def get_color(self, obj_id):
        """Get color of an object"""
        if obj_id in self.objects:
            return self.objects[obj_id][0]
        return None

    def set_color(self, obj_id, color):
        """Set color of an object"""
        if obj_id in self.objects:
            old_color, refs = self.objects[obj_id]
            self.objects[obj_id] = (color, refs)

    def dijkstra_write_barrier(self, from_id, to_id):
        """Dijkstra write barrier"""
        # TODO: Implement Dijkstra write barrier
        # If a black object points to a white object, shade the white object gray
        # This maintains the tricolor invariant: no black object points to white
        pass

    def steele_write_barrier(self, from_id, to_id):
        """Steele write barrier"""
        # TODO: Implement Steele write barrier
        # Shade the source object gray (more conservative than Dijkstra)
        pass

    def generational_write_barrier(self, from_id, to_id):
        """Generational write barrier"""
        # TODO: Implement generational write barrier
        # Record old-to-young pointers in remembered set
        # (Simplified: just record all cross-generation pointers)
        pass

    def write_reference(self, from_id, new_ref, barrier_type="dijkstra"):
        """Update reference with specified write barrier"""
        # TODO: Update reference with specified write barrier
        # 1. Apply write barrier
        # 2. Update the reference
        pass

    def object_count(self):
        """Return number of objects"""
        return len(self.objects)

    def remembered_set_size(self):
        """Return size of remembered set"""
        return len(self.remembered_set)


import unittest


class TestWriteBarrier(unittest.TestCase):
    def test_dijkstra_barrier(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.set_color(obj1, Color.BLACK)
        gc.set_color(obj2, Color.WHITE)

        gc.dijkstra_write_barrier(obj1, obj2)

        # obj2 should now be gray
        self.assertEqual(gc.get_color(obj2), Color.GRAY)

    def test_steele_barrier(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.set_color(obj1, Color.BLACK)

        gc.steele_write_barrier(obj1, obj2)

        # obj1 should now be gray
        self.assertEqual(gc.get_color(obj1), Color.GRAY)

    def test_generational_barrier(self):
        gc = WriteBarrierGC()

        old_obj = gc.allocate()
        young_obj = gc.allocate()

        gc.generational_write_barrier(old_obj, young_obj)

        self.assertEqual(gc.remembered_set_size(), 1)

    def test_write_with_barrier(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.set_color(obj1, Color.BLACK)
        gc.set_color(obj2, Color.WHITE)

        gc.write_reference(obj1, obj2, "dijkstra")

        self.assertEqual(gc.get_color(obj2), Color.GRAY)

    def test_dijkstra_no_change_if_not_white(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.set_color(obj1, Color.BLACK)
        gc.set_color(obj2, Color.BLACK)

        gc.dijkstra_write_barrier(obj1, obj2)

        # obj2 should remain black
        self.assertEqual(gc.get_color(obj2), Color.BLACK)

    def test_steele_always_grays_source(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.set_color(obj1, Color.BLACK)
        gc.set_color(obj2, Color.BLACK)

        gc.steele_write_barrier(obj1, obj2)

        # obj1 should be gray regardless of obj2's color
        self.assertEqual(gc.get_color(obj1), Color.GRAY)

    def test_multiple_generational_barriers(self):
        gc = WriteBarrierGC()

        old1 = gc.allocate()
        old2 = gc.allocate()
        young = gc.allocate()

        gc.generational_write_barrier(old1, young)
        gc.generational_write_barrier(old2, young)

        self.assertEqual(gc.remembered_set_size(), 2)

    def test_write_with_steele_barrier(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.set_color(obj1, Color.BLACK)

        gc.write_reference(obj1, obj2, "steele")

        self.assertEqual(gc.get_color(obj1), Color.GRAY)

    def test_write_with_generational_barrier(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.write_reference(obj1, obj2, "generational")

        self.assertGreaterEqual(gc.remembered_set_size(), 0)

    def test_black_to_gray_dijkstra(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.set_color(obj1, Color.BLACK)
        gc.set_color(obj2, Color.GRAY)

        gc.dijkstra_write_barrier(obj1, obj2)

        # obj2 should remain gray (no change needed)
        self.assertEqual(gc.get_color(obj2), Color.GRAY)

    def test_white_to_white_dijkstra(self):
        gc = WriteBarrierGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.set_color(obj1, Color.WHITE)
        gc.set_color(obj2, Color.WHITE)

        gc.dijkstra_write_barrier(obj1, obj2)

        # No change (source not black)
        self.assertEqual(gc.get_color(obj2), Color.WHITE)


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/01_gc_and_memory/gc08_write_barrier.rs
// gc08_write_barrier.rs
//
// Write barriers intercept pointer writes to maintain GC invariants.
// They're essential for incremental and concurrent GC.
//
// Your task: Implement different write barrier strategies.

// I AM NOT DONE

use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct ObjectId(usize);

#[derive(Debug, Clone, Copy, PartialEq)]
enum Color {
    White,
    Gray,
    Black,
}

pub struct WriteBarrierGC {
    objects: HashMap<ObjectId, (Color, Vec<ObjectId>)>,
    next_id: usize,
    roots: HashSet<ObjectId>,
    remembered_set: HashSet<(ObjectId, ObjectId)>,
}

impl WriteBarrierGC {
    pub fn new() -> Self {
        Self {
            objects: HashMap::new(),
            next_id: 0,
            roots: HashSet::new(),
            remembered_set: HashSet::new(),
        }
    }

    pub fn allocate(&mut self, references: Vec<ObjectId>) -> ObjectId {
        let id = ObjectId(self.next_id);
        self.next_id += 1;
        self.objects.insert(id, (Color::White, references));
        id
    }

    pub fn add_root(&mut self, id: ObjectId) {
        self.roots.insert(id);
    }

    fn get_color(&self, id: ObjectId) -> Option<Color> {
        self.objects.get(&id).map(|(color, _)| *color)
    }

    fn set_color(&mut self, id: ObjectId, color: Color) {
        if let Some((obj_color, _)) = self.objects.get_mut(&id) {
            *obj_color = color;
        }
    }

    pub fn dijkstra_write_barrier(&mut self, from: ObjectId, to: ObjectId) {
        // TODO: Implement Dijkstra write barrier
        // If a black object points to a white object, shade the white object gray
        // This maintains the tricolor invariant: no black object points to white
        todo!()
    }

    pub fn steele_write_barrier(&mut self, from: ObjectId, _to: ObjectId) {
        // TODO: Implement Steele write barrier
        // Shade the source object gray (more conservative than Dijkstra)
        todo!()
    }

    pub fn generational_write_barrier(&mut self, from: ObjectId, to: ObjectId) {
        // TODO: Implement generational write barrier
        // Record old-to-young pointers in remembered set
        // (Simplified: just record all cross-generation pointers)
        todo!()
    }

    pub fn write_reference(&mut self, from: ObjectId, new_ref: ObjectId, barrier_type: &str) {
        // TODO: Update reference with specified write barrier
        // 1. Apply write barrier
        // 2. Update the reference
        todo!()
    }

    pub fn object_count(&self) -> usize {
        self.objects.len()
    }

    pub fn remembered_set_size(&self) -> usize {
        self.remembered_set.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_dijkstra_barrier() {
        let mut gc = WriteBarrierGC::new();

        let obj1 = gc.allocate(vec![]);
        let obj2 = gc.allocate(vec![]);

        gc.set_color(obj1, Color::Black);
        gc.set_color(obj2, Color::White);

        gc.dijkstra_write_barrier(obj1, obj2);

        // obj2 should now be gray
        assert_eq!(gc.get_color(obj2), Some(Color::Gray));
    }

    #[test]
    fn test_steele_barrier() {
        let mut gc = WriteBarrierGC::new();

        let obj1 = gc.allocate(vec![]);
        let obj2 = gc.allocate(vec![]);

        gc.set_color(obj1, Color::Black);

        gc.steele_write_barrier(obj1, obj2);

        // obj1 should now be gray
        assert_eq!(gc.get_color(obj1), Some(Color::Gray));
    }

    #[test]
    fn test_generational_barrier() {
        let mut gc = WriteBarrierGC::new();

        let old_obj = gc.allocate(vec![]);
        let young_obj = gc.allocate(vec![]);

        gc.generational_write_barrier(old_obj, young_obj);

        assert_eq!(gc.remembered_set_size(), 1);
    }

    #[test]
    fn test_write_with_barrier() {
        let mut gc = WriteBarrierGC::new();

        let obj1 = gc.allocate(vec![]);
        let obj2 = gc.allocate(vec![]);

        gc.set_color(obj1, Color::Black);
        gc.set_color(obj2, Color::White);

        gc.write_reference(obj1, obj2, "dijkstra");

        assert_eq!(gc.get_color(obj2), Some(Color::Gray));
    }
}
